sdks/java/nexmark/src/main/java/org/apache/beam/sdk/nexmark/model/sql/adapter/ModelFieldsAdapter.java

```diff
@@ -37,4 +43,19 @@ public RowType getRowType() {
   }
 
   public abstract List<Object> getFieldsValues(T model);
+
+  public abstract T getRowModel(Row row);
+  public abstract Coder getCoder();
+
+  public PCollection<T> applyModel(PCollection<Row> rows) {
```

Comment:
This is essentially the `PTransform.expand()` method.  

I personally think that it would be clearer if instead of applying the `ParDo`, this method would just return the `PTransform` which would be applied somewhere else, e.g. `ModelFieldsAdapter.toModel(Person.getClass())`. Similar to `Combine.perKey()`, `Window.into()`. This way the pipeline is assembled from `PTransforms` in more declarative style, instead of opaquely modifying the `PCollection`.

And it doesn't feel like all queries should expose `applyModel(PCollection)` just for the sake of `NexmarkSqlQuery` calling it. I think that queries either should explicitly handle the conversion internally (copy-pasting) without exposing this method, or don't handle it at all and leave it to `NexmarkSqlQuery` calling the utility methods.

E.g. I think that `if (outModel == null) {` logic from `NexmarkSqlTransform` can be kept in `NexmarkSqlQuery` or implemented as `ModelFieldsAdapter.toModel()`, without having to introduce the `NexmarkSqlTransform` class and extra

Reply:
I was having difficulty using a `PTransform` without definite types. Might just be my lack of java knowledge. I think once I get all the existing SQL queries moved to this model we can remove this and `NexmarkSqlTransform` and just transform to the java pojo in `expand`. I could do that all in one PR, but it would be rather large.

Reply:
Cool, I'm fine with it

Reply:
I agree with what @akedin said about `applyModel`. I like the refactoring you did and the removal of `NexmarkSqlTransform`